ffmpeg: improve cross platform support & fix licensing issues, resolves #7013

codyopel opened the pull request:
- Removed asserts for (l)gpl & (l)gpl3 making it easier to enable
  and disable all (l)gpl & (l)gpl3 portions
- Fix build inputs so that nonfree software isn't pulled in when building ffmpeg
- Fix compiler on systems with clang
- Disable libraries that fail on darwin until they can be fixed
- Fix linux only build inputs so that they aren't pulled in on other platforms
- Fix optional build inputs
- Opencv causes circular dependency issues so it is disabled until it can be resolved

spwhitt commented:
Apparently checking if a package is null causes it to evaluate. For instance, the following line:

```
(mkFlag (v4l_utils != null && isLinux) "0.9" "libv4l2")
```

contains `v4l_utils != null`. This evaluates the `v4l_utils` package, which results in an error on Darwin since this package is not supported. The easiest way to fix this is to swap the arguments to `&&`:

```
(mkFlag (isLinux && v4l_utils != null) "0.9" "libv4l2")
```

but this is potentially brittle.

henrytill commented on pkgs/development/libraries/ffmpeg/generic.nix:
```diff
@@ -449,14 +444,18 @@ stdenv.mkDerivation rec {
-    zeromq4 zlib
-  ] ++ optional faacExtlib faac;
+    bzip2 celt fontconfig freetype fribidi game-music-emu gnutls ladspaH
+    lame libass libbluray libbs2b libcaca libdc1394 libmodplug libogg
```
`libmodplug` isn't enabled on Darwin yet.

henrytill commented:
Confirming @spwhitt's experience:

```bash 
$ nix-build -A ffmpeg
error: Package ‘frei0r-plugins-1.4’ in ‘/Users/ht/src/nixpkgs/pkgs/development/libraries/frei0r/default.nix:16’ is not supported on ‘x86_64-darwin’, refusing to evaluate.
```

codyopel commented:
@spwhitt @henrytill 
Thanks for testing it.  And yeah it wouldn't be ideal to just put ```isLinux``` before other arguments so I'll look into it and find a cleaner way to do it.  It will probably be a bit more complicated to do it cleanly, but that is ok.  I will also disable libmodplug for darwin and add it to the list of things to fix.

codyopel commented:
I updated the build with a function ```disDarwinFix``` for the temporary hacks to disable broken dependencies and for permanent fixes I wrapped the configure flags arguments in an if statement so that is isn't to complicated to modify (```disDarwinFix``` does the same thing, but is cleaner so that it is easier to remove once things are fixed).  Hopefully this should resolve the issue if it failing to evaluate because of the unsupported dependencies in the arguments.

henrytill commented:
I get this:

```
$ nix-build -A ffmpeg
error: while evaluating the attribute ‘configureFlags’ of the derivation ‘ffmpeg-2.5.4’ at /Users/ht/src/nixpkgs/pkgs/development/libraries/ffmpeg/generic.nix:276:3:
while evaluating ‘verFix’ at /Users/ht/src/nixpkgs/pkgs/development/libraries/ffmpeg/generic.nix:206:32, called from /Users/ht/src/nixpkgs/pkgs/development/libraries/ffmpeg/generic.nix:372:6:
attempt to call something which is not a function but a boolean, at /Users/ht/src/nixpkgs/pkgs/development/libraries/ffmpeg/generic.nix:372:14
```

codyopel commented:
@henrytill 
Sorry, that is a typo on line 372.  It should be:
```nix
(verFix (mkFlag (disDarwinFix (gsm != null)) "0.5" "libgsm") "0.5" "--disable-libgsm")
```
I'll update it when I get back.

henrytill commented:
So it successfully built with the above changes, but I discovered that the following also need to be disabled (for now) to get functional executables:
* `game-music-emu`
* `openjpeg_1`
* `vid-stab`
* `x265`

codyopel commented:
A work around for libmodplug was added when I updated the commit.

I don't believe that the issue is with those dependencies (gme, openjpeg, vid.stab, x265), this might be related to the recent cmake issues as they all use cmake.  They have all built successfully on the darwin hydra instance (except openjpeg, but I think I know how to fix that too) but the builds only recently are failing due to libc++abi #6942.

henrytill commented:
Yes, all of these additional dependencies build fine on Darwin, but either the proper dylibs for Darwin aren't being built, or they somehow aren't properly linked to the `ffmpeg` and `ffplay` binaries, and they cause them to fail with `*.dylib not found` errors.  If you could disable them for now, I will look into debugging them later along with some of the other dependencies.  

I'm a bit confused by what you are saying regarding `libmodplug`: it wasn't placed in the Darwin optional section of `buildInputs` when you updated it here (see the above diff), which is where it needs to be in order to build correctly.

henrytill commented:
More info: https://gist.github.com/henrytill/728c40c48e750da28f73

(Resulting from applying the `libgsm` and `libmodplug` fixes and `fftw` and `soxr` PRs)

codyopel commented:
Sorry, I added the darwin flag for libmodplug, that's what I thought you meant, I'll fix the build input.  I was wondering if it wasn't cmake and that it was another linking issue, just wasn't sure since they were all using cmake with the recent issues.  I will add these corrections and hopefully it will build successfully now.  I make sure that the other PR's are merged first too since they are dependent.

codyopel commented:
I think all of the necessary changes are there now, let me know if anything else fails after this.  If everything works we also need to test the changes on ```ffmpeg_0_10``` since we build against it too.

henrytill commented:
Fixes shown here for clarity: https://github.com/codyopel/nixpkgs/compare/ffmpeg-darwin...henrytill:ffmpeg-darwin

Just tested  `ffmpeg_0_10` with these fixes, and that seems to build and function as expected, and  the results of `otool -L ./result/bin/*` all look good.

codyopel committed:
ffmpeg: improve cross platform support & fix licensing issues, resolves #7013
+ Removed asserts for (l)gpl & (l)gpl3 making it easier to enable and disable all
  (l)gpl & (l)gpl3 portions
+ Fix build inputs so that nonfree software isn't pulled in when building ffmpeg
+ Fix compiler on systems with clang
+ Disable libraries that fail on darwin until they can be fixed
+ Fix linux only build inputs so that they aren't pulled in on other platforms
+ Fix optional build inputs
+ Opencv causes circular dependency issues so it is disabled until it can be resolved

codyopel commented:
Yeah, you're right, I forgot to add the correction for that typo.  Also removed the game-music-emu duplicate.  Thanks for testing 0.10, I just wanted to make sure it would at least build.

henrytill commented:
:+1: 

Everything here looks a-okay.  Tested both `ffmpeg_2_5` (a.k.a. `ffmpeg`) and `ffmpeg_0_10`.

`fftw` and `soxr` PRs were also just merged, so we're good to go on that front.

vcunat commented on pkgs/development/libraries/ffmpeg/generic.nix:
```diff
@@ -203,7 +208,12 @@ let
+
   isCygwin = stdenv.isCygwin;
+  isDarwin = stdenv.isDarwin;
+  isLinux = stdenv.isLinux;
```
For future reference: in ``let`` you can instead have ``inherit (stdenv) isCygwin isDarwin isLinux;``.
